doc/sphinx/proof-engine/tactics.rst

```diff
@@ -466,52 +466,87 @@ Examples:
-If no numbers are given for hypothesis :token:`ident`, then all the
-occurrences of :token:`term` in the hypothesis are selected. If numbers are
-given, they refer to occurrences of :token:`term` when the term is printed
-using the :flag:`Printing All` flag, counting from left to right. In particular,
-occurrences of :token:`term` in implicit arguments
-(see :ref:`ImplicitArguments`) or coercions (see :ref:`Coercions`) are
-counted.
-
-If a minus sign is given between ``at`` and the list of occurrences, it
-negates the condition so that the clause denotes all the occurrences
-except the ones explicitly mentioned after the minus sign.
-
-As an exception to the left-to-right order, the occurrences in
-the return subexpression of a match are considered *before* the
-occurrences in the matched term.
-
-In the second case, the ``*`` on the left of ``|-`` means that all occurrences
-of term are selected in every hypothesis.
-
-In the first and second case, if ``*`` is mentioned on the right of ``|-``, the
-occurrences of the conclusion of the goal have to be selected. If some numbers
-are given, then only the occurrences denoted by these numbers are selected. If
-no numbers are given, all occurrences of :token:`term` in the goal are selected.
-
-Finally, the last notation is an abbreviation for ``* |- *``. Note also
-that ``|-`` is optional in the first case when no ``*`` is given.
-
-Here are some tactics that understand occurrence clauses: :tacn:`set`,
-:tacn:`remember`, :tacn:`induction`, :tacn:`destruct`.
+   .. insertprodn occurrences concl_occs
+
+   .. prodn::
+      occurrences ::= at @occs_nums
+      | in @goal_occurrences
+      occs_nums ::= {? - } {+ {| @natural | @ident } }
+      goal_occurrences ::= {*, @hyp_occs } {? %|- {? @concl_occs } }
+      | * %|- {? @concl_occs }
+      | * {? at @occs_nums }
+      hyp_occs ::= @hypident {? at @occs_nums }
+      hypident ::= @ident
+      | ( type of @ident )
+      | ( value of @ident )
+      concl_occs ::= * {? at @occs_nums }
+
+   :n:`@occurrences`
+     The first form of :token:`occurrences` selects occurrences in
+     the conclusion of the goal.  The second form can select occurrences
+     in the goal conclusion and in one or more hypotheses.
+
+   :n:`{? - } {+ {| @natural | @ident } }`
+     Selects the specified occurrences within a single goal or hypothesis.
+     Occurrences are numbered from left to right starting with 1 when the
+     goal is printed with the :flag:`Printing All`.  (In particular, occurrences
+     in implicit arguments (see :ref:`ImplicitArguments`) and
+     coercions (see :ref:`Coercions`) are counted but not shown by default.)
+
+     Specifying `-` includes all occurrences *except* the ones listed.
+
+   :n:`{*, @hyp_occs } {? %|- {? @concl_occs } }`
+     Selects occurrences in the specified hypotheses and the
+     specified occurrences in the conclusion.
+
+     .. todo PR: if above is empty it's useless, right?
```

Comment:
If I understood your comment correctly, yes. Writing `in |- * at ...` would be equivalent to writing just `at ...` (but some people could still prefer this form?).

Reply:
My thought is more that "red in." is a bit strange.  I'd expect this means no occurrences are selected, hence a no-op.  But harmless, so don't need to mention in the doc.

Reply:
Ah OK. But then it corresponds to #13235 / #13237, right?

Reply:
Yes.